catalog/plugin/catalog-plugin-configuration-mapping/src/main/java/org/codice/ddf/catalog/plugin/configuration/mapping/ConfigurationMappingPlugin.java

```diff
@@ -60,35 +67,77 @@ public DeleteRequest process(DeleteRequest input)
         return input;
     }
 
-    public void setExpansionService(Expansion expansionService) {
-        this.expansionService = expansionService;
+    private void expandAttributes(List<Metacard> metacards) {
+
+        AttributeRuleSet attributeRuleSet = new AttributeRuleSet(expansionService);
+
+        for (Metacard metacard : metacards) {
+            attributeRuleSet.expandMetacard(metacard);
+        }
     }
 
-    private void expandAttributes(List<Metacard> metacards) {
-        for (Map.Entry<String, List<String[]>> expansion : expansionService.getExpansionMap()
-                .entrySet()) {
-            String attribute = expansion.getKey();
-            Map<String, String> expansionMap = toMap(expansion.getValue());
-
-            for (Metacard metacard : metacards) {
-                Attribute metacardAttribute = metacard.getAttribute(attribute);
-
-                if (metacardAttribute != null
-                        && expansionMap.containsKey(metacardAttribute.getValue())) {
-                    Attribute attributeImpl = new AttributeImpl(attribute,
-                            expansionMap.get(metacardAttribute.getValue()));
-                    metacard.setAttribute(attributeImpl);
+    private class AttributeRuleSet {
+
+        private Map<String, Map<String, String>> attributeRuleSet;
+
+        public AttributeRuleSet(Expansion expansion) {
+
+            attributeRuleSet = new HashMap<>();
+
+            for (Map.Entry<String, List<String[]>> attributeRules : expansion.getExpansionMap()
+                    .entrySet()) {
+                Map<String, String> attributeMap = toMap(attributeRules.getValue());
+
+                attributeRuleSet.put(attributeRules.getKey(), attributeMap);
+            }
+        }
+
+        public void expandMetacard(Metacard metacard) {
+
+            Set<String> attributeSet = metacard.getMetacardType()
+                    .getAttributeDescriptors()
+                    .stream()
+                    .map(des -> des.getName())
+                    .collect(Collectors.toSet());
+
+            Set<String> commonAttributes = new HashSet<>(CollectionUtils.intersection(attributeSet,
+                    attributeRuleSet.keySet()));
+
+            for (String attributeName : commonAttributes) {
+                Attribute metacardAttribute = metacard.getAttribute(attributeName);
+
+                if (metacardAttribute != null) {
+
+                    // expandTo holds the value that the ruleset
+                    String expandTo = getCorrespondingValue(attributeName,
+                            metacardAttribute.getValue());
```

Comment:
Fair, I think one thing we didn't realize before is that the interface actually already supports this and we don't have to iterate over it ourselves. I'm currently testing it now.